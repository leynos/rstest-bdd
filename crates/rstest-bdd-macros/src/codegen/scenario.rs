//! Code generation for scenario tests.

use super::keyword_to_token;
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::quote;

/// Create a `LitStr` from an examples table cell.
fn cell_to_lit(value: &str) -> syn::LitStr {
    syn::LitStr::new(value, proc_macro2::Span::call_site())
}

/// Generate attributes for rstest cases based on examples.
pub(crate) fn generate_case_attrs(
    examples: &crate::parsing::examples::ExampleTable,
) -> Vec<TokenStream2> {
    examples
        .rows
        .iter()
        .filter(|row| row.iter().any(|cell| !cell.is_empty()))
        .map(|row| {
            let cells = row.iter().map(|v| {
                let lit = cell_to_lit(v);
                quote! { #lit }
            });
            quote! { #[case( #(#cells),* )] }
        })
        .collect()
}

/// Configuration for generating code for a single scenario test.
pub(crate) struct ScenarioConfig<'a> {
    /// Attributes on the annotated function.
    pub(crate) attrs: &'a [syn::Attribute],
    /// Visibility of the function.
    pub(crate) vis: &'a syn::Visibility,
    /// Signature of the function.
    pub(crate) sig: &'a syn::Signature,
    /// Function body.
    pub(crate) block: &'a syn::Block,
    /// Fully qualified feature file path.
    pub(crate) feature_path: String,
    /// Name of the scenario.
    pub(crate) scenario_name: String,
    /// Steps in the scenario.
    pub(crate) steps: Vec<crate::parsing::feature::ParsedStep>,
    /// Examples table for scenario outlines.
    pub(crate) examples: Option<crate::parsing::examples::ExampleTable>,
}

/// Generate tokens representing an optional data table.
///
/// # Examples
/// ```rust,ignore
/// let tokens = generate_table_tokens(None);
/// assert_eq!(tokens.to_string(), "None");
/// ```
fn generate_table_tokens(table: Option<&[Vec<String>]>) -> TokenStream2 {
    table.map_or_else(
        || quote! { None },
        |rows| {
            if rows.is_empty() {
                // Explicitly type the empty slice to avoid inference pitfalls when no rows exist.
                return quote! { Some(&[] as &[&[&str]]) };
            }
            let row_tokens = rows.iter().map(|row| {
                let cells = row.iter().map(|cell| {
                    let lit = cell_to_lit(cell);
                    quote! { #lit }
                });
                quote! { &[#(#cells),*][..] }
            });
            quote! { Some(&[#(#row_tokens),*][..]) }
        },
    )
}

/// Process parsed steps into tokens for keywords, values, and tables.
///
/// # Examples
/// ```rust,ignore
/// use rstest_bdd::StepKeyword;
/// use crate::parsing::feature::ParsedStep;
/// let steps = vec![ParsedStep { keyword: StepKeyword::Given, text: "x".into(), table: None }];
/// let (k, v, t) = process_steps(&steps);
/// assert_eq!(v.len(), 1);
/// ```
fn process_steps(
    steps: &[crate::parsing::feature::ParsedStep],
) -> (Vec<TokenStream2>, Vec<TokenStream2>, Vec<TokenStream2>) {
    let keywords = steps.iter().map(|s| keyword_to_token(s.keyword)).collect();
    let values = steps
        .iter()
        .map(|s| {
            let lit = syn::LitStr::new(&s.text, proc_macro2::Span::call_site());
            quote! { #lit }
        })
        .collect();
    let tables = steps
        .iter()
        .map(|s| generate_table_tokens(s.table.as_deref()))
        .collect();
    (keywords, values, tables)
}

/// Grouped tokens for scenario steps.
struct ProcessedSteps {
    keywords: Vec<TokenStream2>,
    values: Vec<TokenStream2>,
    tables: Vec<TokenStream2>,
}

/// Configuration for generating test tokens.
struct TestTokensConfig<'a> {
    processed_steps: ProcessedSteps,
    feature_path: &'a str,
    scenario_name: &'a str,
    block: &'a syn::Block,
}

/// Generate the inner body of the scenario test.
///
/// # Examples
/// ```rust,ignore
/// # use syn::parse_quote;
/// let processed = ProcessedSteps {
///     keywords: vec![],
///     values: vec![],
///     tables: vec![],
/// };
/// let config = TestTokensConfig {
///     processed_steps: processed,
///     feature_path: "feature",
///     scenario_name: "scenario",
///     block: &parse_quote!({}),
/// };
/// let body = generate_test_tokens(config, std::iter::empty());
/// assert!(body.to_string().contains("StepContext"));
/// ```
fn generate_test_tokens(
    config: TestTokensConfig<'_>,
    ctx_inserts: impl Iterator<Item = TokenStream2>,
) -> TokenStream2 {
    let TestTokensConfig {
        processed_steps:
            ProcessedSteps {
                keywords,
                values,
                tables,
            },
        feature_path,
        scenario_name,
        block,
    } = config;

    quote! {
        let steps = [#((#keywords, #values, #tables)),*];
        let mut ctx = rstest_bdd::StepContext::default();
        #(#ctx_inserts)*
        for (index, (keyword, text, table)) in steps.iter().enumerate() {
            if let Some(f) = rstest_bdd::find_step(*keyword, (*text).into()) {
                if let Err(err) = f(&ctx, *text, *table) {
                    panic!(
                        "Step failed at index {}: {} {} - {}\n(feature: {}, scenario: {})",
                        index,
                        keyword.as_str(),
                        text,
                        err,
                        #feature_path,
                        #scenario_name
                    );
                }
            } else {
                panic!(
                    "Step not found at index {}: {} {} (feature: {}, scenario: {})",
                    index,
                    keyword.as_str(),
                    text,
                    #feature_path,
                    #scenario_name
                );
            }
        }
        #block
    }
}

/// Generate the runtime test for a single scenario.
pub(crate) fn generate_scenario_code(
    config: ScenarioConfig<'_>,
    ctx_inserts: impl Iterator<Item = TokenStream2>,
) -> TokenStream {
    let ScenarioConfig {
        attrs,
        vis,
        sig,
        block,
        feature_path,
        scenario_name,
        steps,
        examples,
    } = config;
    let (keywords, values, tables) = process_steps(&steps);
    let processed_steps = ProcessedSteps {
        keywords,
        values,
        tables,
    };
    let test_config = TestTokensConfig {
        processed_steps,
        feature_path: &feature_path,
        scenario_name: &scenario_name,
        block,
    };
    let case_attrs = examples.map_or_else(Vec::new, |ex| generate_case_attrs(&ex));
    let body = generate_test_tokens(test_config, ctx_inserts);
    TokenStream::from(quote! {
        #[rstest::rstest]
        #(#case_attrs)*
        #(#attrs)*
        #vis #sig { #body }
    })
}
