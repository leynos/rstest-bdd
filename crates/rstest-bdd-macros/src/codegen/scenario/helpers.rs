use proc_macro2::TokenStream as TokenStream2;
use quote::{quote, ToTokens};

/// Create a `LitStr` from an examples table cell.
fn cell_to_lit(value: &str) -> syn::LitStr {
    syn::LitStr::new(value, proc_macro2::Span::call_site())
}

/// Generate attributes for rstest cases based on examples.
pub(crate) fn generate_case_attrs(
    examples: &crate::parsing::examples::ExampleTable,
) -> Vec<TokenStream2> {
    examples
        .rows
        .iter()
        .filter(|row| row.iter().any(|cell| !cell.is_empty()))
        .map(|row| {
            let cells = row.iter().map(|v| {
                let lit = cell_to_lit(v);
                quote! { #lit }
            });
            quote! { #[case( #(#cells),* )] }
        })
        .collect()
}

fn generate_table_tokens(table: Option<&[Vec<String>]>) -> TokenStream2 {
    table.map_or_else(
        || quote! { None },
        |rows| {
            if rows.is_empty() {
                // Explicitly type the empty slice to avoid inference pitfalls when no rows exist.
                quote! { Some(&[] as &[&[&str]]) }
            } else {
                let row_tokens = rows.iter().map(|row| {
                    let cells = row.iter().map(|cell| {
                        let lit = cell_to_lit(cell);
                        quote! { #lit }
                    });
                    quote! { &[#(#cells),*][..] }
                });
                quote! { Some(&[#(#row_tokens),*][..]) }
            }
        },
    )
}

/// Process parsed steps into tokens for keywords, values, and tables.
///
/// # Examples
/// ```rust,ignore
/// use crate::StepKeyword;
/// use crate::parsing::feature::ParsedStep;
/// // Note: `span` is available only with the `compile-time-validation` feature.
/// let steps = vec![ParsedStep {
///     keyword: StepKeyword::Given,
///     text: "x".into(),
///     docstring: None,
///     table: None,
///     span: proc_macro2::Span::call_site(),
/// }];
/// let (k, v, t) = process_steps(&steps);
/// assert_eq!(v.len(), 1);
/// ```
pub(crate) fn process_steps(
    steps: &[crate::parsing::feature::ParsedStep],
) -> (
    Vec<TokenStream2>,
    Vec<TokenStream2>,
    Vec<TokenStream2>,
    Vec<TokenStream2>,
) {
    // Resolve textual conjunctions (And/But) to the previous primary keyword
    // without depending on the validation module, which is behind an optional
    // feature. We seed with the first primary keyword or Given by default.
    let keyword_tokens = {
        let mut prev = steps
            .iter()
            .find_map(|s| match s.keyword {
                crate::StepKeyword::And | crate::StepKeyword::But => None,
                other => Some(other),
            })
            .or(Some(crate::StepKeyword::Given));
        steps.iter().map(move |s| s.keyword.resolve(&mut prev))
    }
    .map(|kw| kw.to_token_stream())
    .collect::<Vec<_>>();
    debug_assert_eq!(keyword_tokens.len(), steps.len());
    let values = steps
        .iter()
        .map(|s| {
            let lit = cell_to_lit(&s.text);
            quote! { #lit }
        })
        .collect();
    let docstrings = steps
        .iter()
        .map(|s| {
            s.docstring.as_ref().map_or_else(
                || quote! { None },
                |d| {
                    let lit = syn::LitStr::new(d, proc_macro2::Span::call_site());
                    quote! { Some(#lit) }
                },
            )
        })
        .collect();
    let tables = steps
        .iter()
        .map(|s| generate_table_tokens(s.table.as_deref()))
        .collect();
    (keyword_tokens, values, docstrings, tables)
}
