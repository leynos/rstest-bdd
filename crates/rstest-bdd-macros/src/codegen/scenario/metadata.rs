use std::fmt;

/// Path to a feature file on disk.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct FeaturePath(String);

impl FeaturePath {
    pub(crate) fn new(value: impl Into<String>) -> Self {
        Self(value.into())
    }

    pub(crate) fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl From<String> for FeaturePath {
    fn from(value: String) -> Self {
        Self::new(value)
    }
}

impl From<&str> for FeaturePath {
    fn from(value: &str) -> Self {
        Self::new(value)
    }
}

impl AsRef<str> for FeaturePath {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl fmt::Display for FeaturePath {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Name of an individual Gherkin scenario.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(crate) struct ScenarioName(String);

impl ScenarioName {
    pub(crate) fn new(value: impl Into<String>) -> Self {
        Self(value.into())
    }

    pub(crate) fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl From<String> for ScenarioName {
    fn from(value: String) -> Self {
        Self::new(value)
    }
}

impl From<&str> for ScenarioName {
    fn from(value: &str) -> Self {
        Self::new(value)
    }
}

impl AsRef<str> for ScenarioName {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl fmt::Display for ScenarioName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}
