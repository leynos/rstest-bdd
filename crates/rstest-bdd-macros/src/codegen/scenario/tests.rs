use super::*;
use crate::parsing::feature::ParsedStep;

#[expect(clippy::expect_used, reason = "test helper with descriptive failures")]
fn kw(ts: &TokenStream2) -> crate::StepKeyword {
    let path = syn::parse2::<syn::Path>(ts.clone()).expect("keyword path");
    let ident = path.segments.last().expect("last").ident.to_string();
    crate::StepKeyword::try_from(ident.as_str()).expect("valid step keyword")
}

fn blank() -> ParsedStep {
    ParsedStep {
        keyword: crate::StepKeyword::Given,
        text: String::new(),
        docstring: None,
        table: None,
        #[cfg(feature = "compile-time-validation")]
        span: proc_macro2::Span::call_site(),
    }
}

#[rstest::rstest]
#[case::leading_and(
    vec![crate::StepKeyword::And, crate::StepKeyword::Then],
    vec![crate::StepKeyword::Then, crate::StepKeyword::Then],
)]
#[case::leading_but(
    vec![crate::StepKeyword::But, crate::StepKeyword::Then],
    vec![crate::StepKeyword::Then, crate::StepKeyword::Then],
)]
#[case::mixed(
    vec![crate::StepKeyword::Given, crate::StepKeyword::And, crate::StepKeyword::But, crate::StepKeyword::Then],
    vec![crate::StepKeyword::Given, crate::StepKeyword::Given, crate::StepKeyword::Given, crate::StepKeyword::Then],
)]
#[case::all_conjunctions(
    vec![crate::StepKeyword::And, crate::StepKeyword::But, crate::StepKeyword::And],
    vec![crate::StepKeyword::Given, crate::StepKeyword::Given, crate::StepKeyword::Given],
)]
#[case::empty(vec![], vec![])]
fn normalises_sequences(
    #[case] seq: Vec<crate::StepKeyword>,
    #[case] expect: Vec<crate::StepKeyword>,
) {
    let steps: Vec<_> = seq
        .into_iter()
        .map(|k| ParsedStep {
            keyword: k,
            ..blank()
        })
        .collect();
    let (keyword_tokens, _, _, _) = process_steps(&steps);
    let parsed: Vec<_> = keyword_tokens.iter().map(kw).collect();
    assert_eq!(parsed, expect);
}

fn tags(list: &[&str]) -> Vec<String> {
    list.iter().map(|tag| (*tag).to_owned()).collect()
}

#[rstest::rstest]
#[case::present(tags(&["@allow_skipped", "@other"]), true)]
#[case::absent(tags(&["@other", "@allow-skip"]), false)]
#[case::empty(Vec::<String>::new(), false)]
#[case::case_sensitive(tags(&["@Allow_Skipped"]), false)]
fn detects_allow_skipped_tag(#[case] tags: Vec<String>, #[case] expected: bool) {
    assert_eq!(scenario_allows_skip(&tags), expected);
}
