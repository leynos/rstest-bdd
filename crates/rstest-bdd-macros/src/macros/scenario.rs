//! Implementation of the `#[scenario]` macro.
//! Binds tests to Gherkin scenarios and validates steps when compile-time flags enable it.

use cfg_if::cfg_if;
use proc_macro::TokenStream;
use std::path::{Path, PathBuf};

use crate::codegen::scenario::{ScenarioConfig, generate_scenario_code};
use crate::parsing::feature::{ScenarioData, extract_scenario_steps, parse_and_load_feature};
use crate::utils::fixtures::extract_function_fixtures;
use crate::validation::parameters::process_scenario_outline_examples;

use syn::{
    LitInt, LitStr,
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    token::Comma,
};

struct ScenarioArgs {
    path: LitStr,
    index: Option<usize>,
}

enum ScenarioArg {
    Path(LitStr),
    Index(usize),
}

impl Parse for ScenarioArg {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        if input.peek(LitStr) {
            let lit: LitStr = input.parse()?;
            Ok(Self::Path(lit))
        } else {
            let ident: syn::Ident = input.parse()?;
            input.parse::<syn::token::Eq>()?;
            if ident == "path" {
                Ok(Self::Path(input.parse()?))
            } else if ident == "index" {
                let li: LitInt = input.parse()?;
                Ok(Self::Index(li.base10_parse()?))
            } else {
                Err(input.error("expected `path` or `index`"))
            }
        }
    }
}

impl Parse for ScenarioArgs {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let args = Punctuated::<ScenarioArg, Comma>::parse_terminated(input)?;
        let mut path = None;
        let mut index = None;

        for arg in args {
            match arg {
                ScenarioArg::Path(lit) => {
                    if path.is_some() {
                        return Err(input.error("duplicate `path` argument"));
                    }
                    path = Some(lit);
                }
                ScenarioArg::Index(i) => {
                    if index.is_some() {
                        return Err(input.error("duplicate `index` argument"));
                    }
                    index = Some(i);
                }
            }
        }

        let path = path.ok_or_else(|| input.error("`path` is required"))?;

        Ok(Self { path, index })
    }
}

pub(crate) fn scenario(attr: TokenStream, item: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(attr as ScenarioArgs);
    let item_fn = syn::parse_macro_input!(item as syn::ItemFn);
    match try_scenario(args, item_fn) {
        Ok(tokens) => tokens,
        Err(err) => err,
    }
}

fn try_scenario(
    ScenarioArgs { path, index }: ScenarioArgs,
    mut item_fn: syn::ItemFn,
) -> std::result::Result<TokenStream, TokenStream> {
    let path = PathBuf::from(path.value());
    let attrs = &item_fn.attrs;
    let vis = &item_fn.vis;
    let sig = &mut item_fn.sig;
    let block = &item_fn.block;

    let feature = parse_and_load_feature(&path).map_err(proc_macro::TokenStream::from)?;
    let feature_path_str = canonical_feature_path(&path);
    let ScenarioData {
        name: scenario_name,
        steps,
        examples,
    } = extract_scenario_steps(&feature, index).map_err(proc_macro::TokenStream::from)?;

    if let Some(err) = validate_steps_compile_time(&steps) {
        return Err(err);
    }

    process_scenario_outline_examples(sig, examples.as_ref())
        .map_err(proc_macro::TokenStream::from)?;

    let (_args, ctx_inserts) = extract_function_fixtures(sig);

    Ok(generate_scenario_code(
        ScenarioConfig {
            attrs,
            vis,
            sig,
            block,
            feature_path: feature_path_str,
            scenario_name,
            steps,
            examples,
        },
        ctx_inserts,
    ))
}

/// Canonicalise the feature path for stable diagnostics.
///
/// ```rust,ignore
/// # use std::path::{Path, PathBuf};
/// # fn demo() {
/// let path = PathBuf::from("features/example.feature");
/// let _ = canonical_feature_path(&path);
/// # }
/// ```
fn canonical_feature_path(path: &Path) -> String {
    std::env::var("CARGO_MANIFEST_DIR")
        .ok()
        .map(PathBuf::from)
        .map(|d| d.join(path))
        .and_then(|p| std::fs::canonicalize(&p).ok())
        .unwrap_or_else(|| PathBuf::from(path))
        .display()
        .to_string()
}

/// Validate registered steps when compile-time validation is enabled.
///
/// ```rust,ignore,ignore
/// let steps = Vec::new();
/// let _ = validate_steps_compile_time(&steps);
/// ```
fn validate_steps_compile_time(
    steps: &[crate::parsing::feature::ParsedStep],
) -> Option<TokenStream> {
    // When both features are enabled, strict mode wins.
    cfg_if! {
        if #[cfg(feature = "strict-compile-time-validation")] {
            crate::validation::steps::validate_steps_exist(steps, true)
                .err()
                .map(|e| e.into_compile_error().into())
        } else if #[cfg(feature = "compile-time-validation")] {
            crate::validation::steps::validate_steps_exist(steps, false)
                .err()
                .map(|e| e.into_compile_error().into())
        } else {
            None
        }
    }
}
#[cfg(test)]
mod tests {
    use super::canonical_feature_path;
    use rstest::rstest;
    use std::env;
    use std::path::{Path, PathBuf};

    #[rstest]
    fn canonicalises_with_manifest_dir() {
        let manifest = match env::var("CARGO_MANIFEST_DIR") {
            Ok(m) => PathBuf::from(m),
            Err(e) => panic!("manifest dir: {e}"),
        };
        let path = Path::new("Cargo.toml");
        let expected = match manifest.join(path).canonicalize() {
            Ok(p) => p.display().to_string(),
            Err(e) => panic!("canonical path: {e}"),
        };
        assert_eq!(canonical_feature_path(path), expected);
    }

    #[rstest]
    fn falls_back_on_missing_path() {
        let path = Path::new("does-not-exist.feature");
        assert_eq!(canonical_feature_path(path), path.display().to_string());
    }
}
