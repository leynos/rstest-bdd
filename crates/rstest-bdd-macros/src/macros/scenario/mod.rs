//! Implements the `#[scenario]` macro, wiring Rust tests to Gherkin scenarios
//! and surfacing compile-time diagnostics for invalid configuration. Supports
//! mutually exclusive selectors that either bind by index or match the
//! case-sensitive scenario title, defaulting to the first scenario when no
//! selector is supplied. An optional `tags = "â€¦"` argument filters candidate
//! scenarios before selector resolution, keeping tests focused on the relevant
//! examples.
//!
//! Tag expressions combine case-sensitive tag names with the operators `not`,
//! `and`, and `or`. The parser applies the precedence `not` > `and` > `or`, and
//! parentheses may override the default binding. For instance, the following
//! test only executes scenarios tagged `@fast` while excluding any marked as
//! `@wip` or `@flaky`:
//!
//! ```ignore
//! #[scenario(
//!     "tests/features/filtering.feature",
//!     tags = "@fast and not (@wip or @flaky)"
//! )]
//! fn fast_stable_cases() {}
//! ```

mod args;
mod paths;

use proc_macro::TokenStream;
use proc_macro2::Span;
use std::path::PathBuf;

use crate::codegen::scenario::{generate_scenario_code, FeaturePath, ScenarioConfig, ScenarioName};
use crate::parsing::feature::{extract_scenario_steps, parse_and_load_feature, ScenarioData};
use crate::parsing::tags::TagExpression;
use crate::utils::fixtures::extract_function_fixtures;
use crate::validation::parameters::process_scenario_outline_examples;

use self::args::{ScenarioArgs, ScenarioSelector};
use self::paths::canonical_feature_path;

struct ScenarioTagFilter {
    expr: TagExpression,
    span: Span,
    raw: String,
}

/// Encapsulates the data needed to search and filter scenarios.
#[derive(Copy, Clone)]
struct ScenarioLookup<'a> {
    feature: &'a gherkin::Feature,
    candidate_indices: &'a [usize],
    tag_filter: Option<&'a ScenarioTagFilter>,
}

pub(crate) fn scenario(attr: TokenStream, item: TokenStream) -> TokenStream {
    let args = syn::parse_macro_input!(attr as ScenarioArgs);
    let item_fn = syn::parse_macro_input!(item as syn::ItemFn);
    match try_scenario(args, item_fn) {
        Ok(tokens) => tokens,
        Err(err) => err,
    }
}

fn try_scenario(
    ScenarioArgs {
        path,
        selector,
        tag_filter,
    }: ScenarioArgs,
    mut item_fn: syn::ItemFn,
) -> std::result::Result<TokenStream, TokenStream> {
    let path_lit = path;
    let path = PathBuf::from(path_lit.value());
    let attrs = &item_fn.attrs;
    let vis = &item_fn.vis;
    let sig = &mut item_fn.sig;
    let block = &item_fn.block;

    // Retrieve cached feature to avoid repeated parsing.
    let feature = parse_and_load_feature(&path).map_err(proc_macro::TokenStream::from)?;
    let tag_filter = parse_tag_filter(tag_filter)?;
    ensure_feature_not_empty(&path_lit, &feature)?;
    let candidate_indices = resolve_candidate_indices(selector.as_ref(), &feature, &path_lit)?;
    let scenario_data = select_scenario(
        ScenarioLookup {
            feature: &feature,
            candidate_indices: &candidate_indices,
            tag_filter: tag_filter.as_ref(),
        },
        selector.as_ref(),
        &path_lit,
    )?;

    let feature_path_str = canonical_feature_path(&path);
    let ScenarioData {
        name: scenario_name,
        steps,
        examples,
        tags,
    } = scenario_data;
    let allow_skipped = crate::codegen::scenario::scenario_allows_skip(&tags);

    if let Some(err) = validate_steps_compile_time(&steps) {
        return Err(err);
    }

    process_scenario_outline_examples(sig, examples.as_ref())
        .map_err(proc_macro::TokenStream::from)?;

    let (_args, ctx_inserts) = extract_function_fixtures(sig)
        .map_err(|err| proc_macro::TokenStream::from(err.into_compile_error()))?;

    Ok(generate_scenario_code(
        ScenarioConfig {
            attrs,
            vis,
            sig,
            block,
            feature_path: FeaturePath::new(feature_path_str),
            scenario_name: ScenarioName::new(scenario_name),
            steps,
            examples,
            allow_skipped,
        },
        ctx_inserts.into_iter(),
    ))
}

fn parse_tag_filter(
    tag_filter: Option<syn::LitStr>,
) -> Result<Option<ScenarioTagFilter>, TokenStream> {
    tag_filter
        .map(|lit| {
            let raw = lit.value();
            TagExpression::parse(&raw)
                .map(|expr| ScenarioTagFilter {
                    expr,
                    span: lit.span(),
                    raw,
                })
                .map_err(|err| {
                    proc_macro::TokenStream::from(
                        syn::Error::new(lit.span(), err.to_string()).into_compile_error(),
                    )
                })
        })
        .transpose()
}

fn ensure_feature_not_empty(
    path_lit: &syn::LitStr,
    feature: &gherkin::Feature,
) -> Result<(), TokenStream> {
    if feature.scenarios.is_empty() {
        let msg = format!("feature `{}` contains no scenarios", path_lit.value());
        let err = syn::Error::new(path_lit.span(), msg);
        Err(proc_macro::TokenStream::from(err.into_compile_error()))
    } else {
        Ok(())
    }
}

fn resolve_candidate_indices(
    selector: Option<&ScenarioSelector>,
    feature: &gherkin::Feature,
    path_lit: &syn::LitStr,
) -> Result<Vec<usize>, TokenStream> {
    match selector {
        Some(ScenarioSelector::Index { value, span }) => {
            let value = *value;
            if value >= feature.scenarios.len() {
                let count = feature.scenarios.len();
                let noun = if count == 1 { "scenario" } else { "scenarios" };
                let message = format!(
                    "scenario index {value} out of range; feature `{}` defines {count} {noun}",
                    path_lit.value()
                );
                let err = syn::Error::new(*span, message);
                Err(proc_macro::TokenStream::from(err.into_compile_error()))
            } else {
                Ok(vec![value])
            }
        }
        Some(ScenarioSelector::Name { value, span }) => {
            let idx = find_scenario_by_name(feature, value, *span)
                .map_err(|err| proc_macro::TokenStream::from(err.into_compile_error()))?;
            Ok(vec![idx])
        }
        None => Ok((0..feature.scenarios.len()).collect()),
    }
}

fn select_scenario(
    lookup: ScenarioLookup<'_>,
    selector: Option<&ScenarioSelector>,
    path_lit: &syn::LitStr,
) -> Result<ScenarioData, TokenStream> {
    let mut examined_tag_sets: Vec<Vec<String>> = Vec::new();

    for &idx in lookup.candidate_indices {
        let mut data = extract_scenario_steps(lookup.feature, Some(idx))
            .map_err(proc_macro::TokenStream::from)?;
        let matches = lookup
            .tag_filter
            .map_or(true, |filter| data.filter_by_tags(&filter.expr));
        if matches {
            return Ok(data);
        }
        if lookup.tag_filter.is_some() {
            // Preserve the feature + scenario tags that callers evaluated so the
            // eventual diagnostic can explain which annotations were available.
            examined_tag_sets.push(data.tags.clone());
        }
    }

    lookup.tag_filter.map_or_else(
        || {
            debug_assert!(
                lookup.candidate_indices.is_empty(),
                "expected default scenario selection to succeed when no tag filter is provided",
            );
            // This branch should be unreachable in practice because
            // `ensure_feature_not_empty` guards empty features. Emit a specific
            // diagnostic if it ever triggers so consumers know the macro input
            // lacked scenarios.
            let message = format!("feature `{}` contains no scenarios", path_lit.value());
            let err = syn::Error::new(path_lit.span(), message);
            Err(proc_macro::TokenStream::from(err.into_compile_error()))
        },
        |filter| {
            let available_clause = format_available_tags(&examined_tag_sets);
            let message = match selector {
                Some(ScenarioSelector::Index { value, .. }) => {
                    format!(
                        "scenario at index {} does not match tag expression `{}`; {}",
                        value, filter.raw, available_clause
                    )
                }
                Some(ScenarioSelector::Name { value, .. }) => {
                    format!(
                        "scenario named \"{value}\" does not match tag expression `{}`; {}",
                        filter.raw, available_clause
                    )
                }
                None => format!(
                    "no scenarios matched tag expression `{}`; {}",
                    filter.raw, available_clause
                ),
            };
            let err = syn::Error::new(filter.span, message);
            Err(proc_macro::TokenStream::from(err.into_compile_error()))
        },
    )
}

fn format_available_tags(tag_sets: &[Vec<String>]) -> String {
    // Multiple scenarios can feed the diagnostic when no selector is supplied;
    // serialise each tag set separately so callers can still spot gaps without
    // losing the original grouping.
    if tag_sets.is_empty() {
        return "available tags: <none>".to_string();
    }

    let formatted_sets: Vec<String> = tag_sets
        .iter()
        .map(|tags| {
            if tags.is_empty() {
                "<none>".to_string()
            } else {
                tags.join(", ")
            }
        })
        .collect();

    format!("available tags: {}", formatted_sets.join("; "))
}

fn find_scenario_by_name(
    feature: &gherkin::Feature,
    name: &str,
    span: Span,
) -> Result<usize, syn::Error> {
    let mut matches = feature
        .scenarios
        .iter()
        .enumerate()
        .filter(|(_, scenario)| scenario.name == name);

    match matches.next() {
        None => {
            let available: Vec<String> = feature
                .scenarios
                .iter()
                .map(|scenario| format!("\"{}\"", scenario.name))
                .collect();
            let message = if available.is_empty() {
                format!("scenario named \"{name}\" not found; feature contains no scenarios")
            } else {
                let options = available.join(", ");
                format!("scenario named \"{name}\" not found; available titles: {options}")
            };
            Err(syn::Error::new(span, message))
        }
        Some((idx, scenario)) => {
            let rest: Vec<(usize, &gherkin::Scenario)> = matches.collect();
            if rest.is_empty() {
                Ok(idx)
            } else {
                let mut all = Vec::with_capacity(rest.len() + 1);
                all.push((idx, scenario));
                all.extend(rest);
                let indexes = all
                    .iter()
                    .map(|(index, _)| index.to_string())
                    .collect::<Vec<_>>()
                    .join(", ");
                let lines = all
                    .iter()
                    .map(|(_, matched_scenario)| matched_scenario.position.line.to_string())
                    .collect::<Vec<_>>()
                    .join(", ");
                let message = format!(
                    "found multiple scenarios named \"{name}\"; use the `index` selector to disambiguate (matching indexes: {indexes}; lines: {lines})",
                );
                Err(syn::Error::new(span, message))
            }
        }
    }
}

fn validate_steps_compile_time(
    steps: &[crate::parsing::feature::ParsedStep],
) -> Option<TokenStream> {
    let res: Result<(), syn::Error> = {
        cfg_if::cfg_if! {
            if #[cfg(feature = "strict-compile-time-validation")] {
                crate::validation::steps::validate_steps_exist(steps, true)
            } else if #[cfg(feature = "compile-time-validation")] {
                crate::validation::steps::validate_steps_exist(steps, false)
            } else {
                let _ = steps;
                Ok(())
            }
        }
    };
    res.err()
        .map(|e| proc_macro::TokenStream::from(e.into_compile_error()))
}
