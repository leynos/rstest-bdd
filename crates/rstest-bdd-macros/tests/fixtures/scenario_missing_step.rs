// compile-flags: --cfg feature="strict-compile-time-validation"
use rstest_bdd_macros::scenario;

#[scenario(path = "../../../../crates/rstest-bdd-macros/tests/features/unmatched.feature")]
fn missing_step() {}

fn main() {}
