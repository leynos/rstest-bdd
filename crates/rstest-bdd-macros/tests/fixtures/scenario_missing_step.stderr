error: No matching step definition found for: Given an undefined step
 --> tests/fixtures/scenario_missing_step.rs:4:1
  |
4 | #[scenario(path = "../../../../crates/rstest-bdd-macros/tests/features/unmatched.feature")]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `scenario` (in Nightly builds, run with -Z macro-backtrace for more info)
